### PR TITLE
vec::IntoIter fixes

### PR DESCRIPTION
**Pull request by main--**
Ran into these issues while tracking down a memory leak.

### CODE REVIEW

**Comment by konsumlamm** [comment]
This fixes #178.

File: src/collections/vec.rs [hunk 1]
```diff
@@ -2341,9 +2341,16 @@ impl<'bump, T: 'bump> DoubleEndedIterator for IntoIter<T> {
...
 
-impl<'bump, T: 'bump> FusedIterator for IntoIter<T> {}
+impl<'bump, T: 'bump> FusedIterator for IntoIter<'bump, T> {}
+
+impl<'bump, T> Drop for IntoIter<'bump, T> {
+    fn drop(&mut self) {
+        // drop all remaining elements
+        self.for_each(drop);
+    }
+}
```
Comment (fitzgen) [inline src/collections/vec.rs:2353]: Can this use a drop guard the same way that `alloc` does?

https://cs.github.com/rust-lang/rust/blob/9b0a099dfc9a97ecd10adb319396c731c4b2d169/library/alloc/src/vec/into_iter.rs?q=drop+for+intoiter#L355-L377

**Review by fitzgen — COMMENTED** [review]
Thanks! One comment to be addressed before this can merge.

File: src/collections/vec.rs (hunk 1 above)
Comment (main--) [reply src/collections/vec.rs:2353]: If my understanding of the implementation you linked is correct, then no. It seems to me that the idea there is that if an element destructor were to panic then at least the Vec’s allocation would still be freed.

This should already work the same way in bumpalo, given that the backing allocation belongs to the Bump struct, which would still be dropped normally during unwinding, if IntoIter’s drop were to panic.

**Review by fitzgen — APPROVED** [review]
Thanks again!

**Pull request merged**